packages/ui/Scrollable/Scrollable.js

```diff
@@ -785,18 +826,57 @@ class ScrollableBase extends Component {
 		forward(type, {scrollLeft: this.scrollLeft, scrollTop: this.scrollTop, moreInfo: this.getMoreInfo()}, this.props);
 	}
 
+	compareAndSet (obj, name, value) {
+		const changed = (obj[name] !== value);
+		obj[name] = value;
+		return changed;
+	}
+
+	forwardReachEdgeEvent (orientation) {
+		const {onReachEdge} = this.props;
+		if (onReachEdge) {
+			const {direction} = this.props;
+
+			if (direction === orientation || direction === 'both') {
+				const
+					{compareAndSet} = this,
+					info = this.touchingEdgeInfo[orientation];
+				let
+					conditionBegin = false,
+					conditionEnd = false;
+
+				if (this.getScrollabilities(orientation)) {
+					const
+						name = (orientation === 'vertical') ? 'Top' : 'Left',
+						curPos = this['scroll' + name];
+					conditionBegin = (curPos <= 0);
+					conditionEnd = (curPos >= this.getScrollBounds()['max' + name]);
+				}
+
+				if (compareAndSet(info, 'begin', conditionBegin)) {
+					onReachEdge({orientation, position: 'begin', reached: info.begin});
+				}
+				if (compareAndSet(info, 'end', conditionEnd)) {
+					onReachEdge({orientation, position: 'end', reached: info.end});
+				}
+			}
+		}
+	}
+
 	// update scroll position
 
 	setScrollLeft (value) {
 		const
 			bounds = this.getScrollBounds(),
 			maxValue = bounds.maxLeft,
-			{type, ratio} = this.getOverscrollStatus('horizontal');
+			orientation = 'horizontal',
+			{type, ratio} = this.getOverscrollStatus(orientation);
 
 		this.scrollLeft = clamp(0, maxValue, value);
+		this.forwardReachEdgeEvent(orientation);
```

Comment:
As we discussed, it would be better to move the code in `forwardReachEdgeEvent` here after cleaning up like the following code.

```
		if (onReachEdge && this.getScrollabilities(orientation)) {
			const
				info = this.touchingEdgeInfo[orientation],
				condition = {begin: this.scrollLeft <= 0, end: this.scrollLeft >= this.getScrollBounds()[maxValue]};

			['begin', 'end'].forEach((edge) => {
				if (info[edge] !== condition[edge]) {
					info[edge] = condition[edge];
					onReachEdge({orientation, position: edge, reached: info[edge]});
				}
			});
		}
```

Reply:
👌